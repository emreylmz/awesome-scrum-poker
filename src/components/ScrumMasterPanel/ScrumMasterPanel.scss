@import "../../styles/colors";
@import "../../styles/metrics";

.scrum-master-panel {
  display: flex;
  flex: 1;
  align-items: center;
  flex-direction: column;
  border: $default-border-width solid $primary-color;
  padding: $default-margin;
  border-radius: $default-border-radius;
  margin: $default-margin;

  &__title {

  }

  .voters-container {
    display: flex;
    align-self: stretch;

    .voter-line {
      display: flex;
      align-items: center;

      .voter-name {
        flex: 2;

        &::after {
          content: ':';
        }
      }

      .voter-point {
        flex: 1;
        text-align: center;
        font-size: 14px;
      }
    }
  }


  .button {
    display: flex;
    align-items: flex-end;
  }

}
