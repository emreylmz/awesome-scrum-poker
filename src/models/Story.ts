export interface Story {
  id: string
  title: string
  point?: number
  status: number
}
