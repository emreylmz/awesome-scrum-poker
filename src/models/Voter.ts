export interface Voter {
  id: string
  name: string
  point?: string
  voted?: boolean
  orderInSprint: number
  isScrumMaster?: boolean
}
