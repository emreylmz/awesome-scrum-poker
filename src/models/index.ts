export * from './Sprint'
export * from './Story'
export * from './Voter'
