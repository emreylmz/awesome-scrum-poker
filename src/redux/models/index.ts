export * from './reduxAction'
export * from './reduxType'
