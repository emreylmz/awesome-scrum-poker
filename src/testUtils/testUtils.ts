export const emptyTestFunc = (): void => void 0
