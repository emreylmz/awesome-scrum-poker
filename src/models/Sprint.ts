import { Story } from './Story'
import { Voter } from './Voter'

export interface Sprint {
  id: string
  name: string
  numberOfVoters: number
  storyList: Story[]
  activeStory: Story
  inviteLink?: string
  voterList: Voter[]
}
