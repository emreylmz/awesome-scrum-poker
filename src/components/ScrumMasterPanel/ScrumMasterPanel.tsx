import React from 'react'
import { connect } from 'react-redux'
import { Sprint, Voter } from '../../models'
import { RootState } from '../../redux/reducers'
import { Render } from '../../types'
import Button from '../Button/Button'
import LabeledTextBox from '../LabeledTextBox/LabeledTextBox'
import './ScrumMasterPanel.scss'

export enum VoteState {
  NotVoted,
  EveryOneVoted,
  AllSamePoint
}

const getStoryTitle = (sprint?: Sprint): string => {
  if (!sprint || !sprint.activeStory) {
    return ''
  }
  return sprint.activeStory.title
}

const getVoters = (sprint?: Sprint, selectedPoint?: string): Voter[] => {
  if (!sprint) {
    return []
  }
  const scrumMaster: Voter | undefined = sprint.voterList && sprint.voterList.find(voter => voter.isScrumMaster)
  if (!scrumMaster) {
    return []
  }

  // scrumMaster.point = selectedPoint

  return [...sprint.voterList]
}

const getVoteState = (sprint?: Sprint): VoteState => {
  if (!sprint) {
    return VoteState.NotVoted
  }

  const { voterList: voters } = sprint
  if (voters.some(voter => voter.point === undefined)) {
    return VoteState.NotVoted
  }
  const currentPoint = voters[0].point
  for (const voter of voters) {
    if (voter.point !== currentPoint) {
      return VoteState.EveryOneVoted
    }
  }
  return VoteState.AllSamePoint
}

interface OwnProps {
  onContinueClick: (finalScore: number) => void
}

interface StateProps {
  storyTitle: string
  voters: Voter[]
  voteState: VoteState
}

type Props = StateProps & OwnProps

interface State {
  finalScore?: string
}

export class ScrumMasterPanel extends React.PureComponent<Props, State> {
  private onContinueClick = (): void => {
    const { voters, voteState, onContinueClick } = this.props
    if (voteState === VoteState.NotVoted) {
      return
    }

    const { finalScore: point } = this.state
    if (!point || point === '?') {
      return
    }
    onContinueClick(+point)
  }

  private onFinalScoreChange = (value?: string): void => {
    this.setState({
      finalScore: value
    })
  }

  private renderVoter(voter: Voter): Render {
    return (
      <div key={voter.name} className={'voter-line'}>
        <span className={'voter-name'}>{voter.name}</span>
        <span className={'voter-point'}>{voter.point || (voter.voted ? 'Voted' : 'Not Voted')}</span>
      </div>
    )
  }

  render(): Render {
    const { storyTitle, voters, voteState } = this.props
    return (
      <div className={'scrum-master-panel'}>
        <div className={'scrum-master-panel__title'}>{`${storyTitle} is active`}</div>
        <div className={'voters-container'}>{voters.map(this.renderVoter)}</div>
        {voteState !== VoteState.NotVoted && (
          <LabeledTextBox label={'Final Score'} numeric onChange={this.onFinalScoreChange} />
        )}

        <Button
          title={`End Voting For ${storyTitle}`}
          disabled={voteState !== VoteState.AllSamePoint}
          onClick={this.onContinueClick}
        />
      </div>
    )
  }
}

const mapStateToProps = (state: RootState): StateProps => {
  return {
    storyTitle: getStoryTitle(state.SprintReducer.sprint),
    voters: getVoters(state.SprintReducer.sprint, state.StoryReducer.selectedPoint),
    voteState: getVoteState(state.SprintReducer.sprint)
  }
}

export type ScrumMasterPanelProps = Props
export default connect(mapStateToProps)(ScrumMasterPanel)
