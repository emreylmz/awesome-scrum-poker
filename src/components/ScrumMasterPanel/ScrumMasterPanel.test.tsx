import { shallow, ShallowWrapper } from 'enzyme'
import React from 'react'
import { Voter } from '../../models'
import '../../testUtils/setupTests'
import { emptyTestFunc } from '../../testUtils/testUtils'
import { ScrumMasterPanel, ScrumMasterPanelProps, VoteState } from './ScrumMasterPanel'

type DoneCallback = jest.DoneCallback

const defaultProps: ScrumMasterPanelProps = {
  onContinueClick: emptyTestFunc,
  storyTitle: 'Test Story',
  voters: [],
  voteState: VoteState.NotVoted
}

const getWrapper = (props?: ScrumMasterPanelProps): ShallowWrapper => {
  return shallow(<ScrumMasterPanel {...defaultProps} {...props} />)
}

describe('Scrum Master Panel UI', (): void => {
  it('should render correctly', (done: DoneCallback): void => {
    const wrapper = getWrapper()
    expect(wrapper).toMatchSnapshot()
    done()
  })

  it('should render voters and points correctly', (done: DoneCallback): void => {
    const voters: Voter[] = [
      {
        id: '1',
        point: '1',
        name: 'Voter 1',
        orderInSprint: 0
      },
      {
        id: '2',
        point: '1',
        name: 'Voter 2',
        orderInSprint: 1
      },
      {
        id: '3',
        point: '1',
        name: 'Voter 3',
        orderInSprint: 2,
        isScrumMaster: true
      }
    ]
    const wrapper = getWrapper({ ...defaultProps, voters })

    const voterWrappers = wrapper.find('.voter-line')
    voterWrappers.map((voterWrapper, index) => {
      const voter = voters[index]
      const voterName = voterWrapper.find('.voter-name')
      const voterPoint = voterWrapper.find('.voter-point')
      expect(voterName.text()).toEqual(voter.name)
      expect(voterPoint.text()).toEqual(voter.point)
    })
    done()
  })

  it('should show scrum master point when the scrum master point given', (done: DoneCallback): void => {
    done()
  })

  it('should show error message when End Voting button clicked if final score not entered', (done: DoneCallback): void => {
    done()
  })

  it('should update voters and points', (done: DoneCallback): void => {
    done()
  })

  it('should not display final score input till each teammate voted', (done: DoneCallback): void => {
    done()
  })
})

describe('Scrum Master Panel Logic', (): void => {
  it('should not call onContinue method when End voting button clicked if final score not entered', (done: DoneCallback): void => {
    done()
  })

  it('should not end voting till each teammate voted', (done: DoneCallback): void => {
    done()
  })
})
