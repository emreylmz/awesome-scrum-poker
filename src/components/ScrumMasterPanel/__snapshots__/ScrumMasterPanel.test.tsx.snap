// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scrum Master Panel UI should render correctly 1`] = `ShallowWrapper {}`;
